report/sql_tasks/python_script.py

```diff
@@ -0,0 +1,66 @@
+import textwrap
+from dataclasses import dataclass, field
+from importlib.util import module_from_spec, spec_from_file_location
+from types import ModuleType
+from typing import Any, Optional
+
+from report.sql_tasks.timer import Timer
+
+
+@dataclass
+class PythonScript:
+    """A class representing a Python script.
+
+    This requires you to define a Python file with a `main(**kwargs)` function
+    which will be passed the connection object.
+    """
+
+    path: str
+    """Full path of the script."""
+
+    timing: Timer = field(default_factory=Timer)
+    """Timer for query execution."""
+
+    module: ModuleType = None
+    """The module loaded from the script."""
+
+    result: Optional[Any] = None
+    """The result of executing `main()` if any."""
+
+    def __post_init__(self):
+        if self.module:
+            return
+
+        # Generate a "spec" for the module
+        spec = spec_from_file_location(f"script_{id(self)}", self.path)
+
+        # Create a module from the spec
+        self.module = module_from_spec(spec)
+
+        # Trigger processing of the module
+        spec.loader.exec_module(self.module)
+
+        if not hasattr(self.module, "main"):
+            raise ModuleNotFoundError(
+                f"Expected to find `main()` function in {self.path}"
+            )
+
+    def execute(self, connection):
```

Comment:
Getting the connection here makes sense, the python script can then be involved in the same transaction.

Reply:
It theoretically can, but I had trouble getting that to work practically? We might want to investigate that more later?